File: thread_pool_example.py
"""
线程池线程数设置示例 (Python)

线程池大小的设置原则：
1. CPU密集型任务：线程数 = CPU核心数 + 1
2. IO密集型任务：线程数 = CPU核心数 * (1 + IO等待时间/CPU计算时间)
3. 混合型任务：需要根据实际情况调整
"""

import os
import time
import concurrent.futures
import threading
from multiprocessing import cpu_count


class ThreadPoolExample:
    """线程池示例类"""
    
    def __init__(self):
        self.cpu_cores = cpu_count()
        print(f"CPU核心数: {self.cpu_cores}")
    
    def cpu_intensive_task(self, n):
        """CPU密集型任务示例"""
        result = sum(i * i for i in range(n))
        return result
    
    def io_intensive_task(self, duration):
        """IO密集型任务示例"""
        time.sleep(duration)  # 模拟IO等待
        return f"IO任务完成，耗时 {duration} 秒"
    
    def cpu_intensive_example(self):
        """CPU密集型任务线程池示例"""
        # CPU密集型：线程数 = CPU核心数 + 1
        thread_pool_size = self.cpu_cores + 1
        
        print(f"\n=== CPU密集型任务示例 ===")
        print(f"线程池大小: {thread_pool_size}")
        
        with concurrent.futures.ThreadPoolExecutor(max_workers=thread_pool_size) as executor:
            # 提交CPU密集型任务
            futures = [
                executor.submit(self.cpu_intensive_task, 1000000)
                for _ in range(10)
            ]
            
            # 获取结果
            results = []
            for future in concurrent.futures.as_completed(futures):
                try:
                    result = future.result()
                    results.append(result)
                    print(f"任务完成，结果: {result}")
                except Exception as e:
                    print(f"任务执行出错: {e}")
    
    def io_intensive_example(self):
        """IO密集型任务线程池示例"""
        # IO密集型：线程数 = CPU核心数 * 2 到 CPU核心数 * 4
        thread_pool_size = self.cpu_cores * 2
        
        print(f"\n=== IO密集型任务示例 ===")
        print(f"线程池大小: {thread_pool_size}")
        
        with concurrent.futures.ThreadPoolExecutor(max_workers=thread_pool_size) as executor:
            # 提交IO密集型任务
            futures = [
                executor.submit(self.io_intensive_task, 1)
                for _ in range(20)
            ]
            
            # 获取结果
            for i, future in enumerate(concurrent.futures.as_completed(futures)):
                try:
                    result = future.result()
                    print(f"任务 {i+1}: {result}")
                except Exception as e:
                    print(f"任务执行出错: {e}")
    
    def mixed_task_example(self):
        """混合型任务示例"""
        # 混合型任务：根据实际情况调整，通常设置为 CPU核心数 * 1.5 到 CPU核心数 * 2
        thread_pool_size = int(self.cpu_cores * 1.5)
        
        print(f"\n=== 混合型任务示例 ===")
        print(f"线程池大小: {thread_pool_size}")
        
        def mixed_task(task_id, task_type):
            """混合型任务"""
            if task_type == 'cpu':
                return self.cpu_intensive_task(500000)
            else:
                return self.io_intensive_task(0.5)
        
        with concurrent.futures.ThreadPoolExecutor(max_workers=thread_pool_size) as executor:
            tasks = []
            for i in range(15):
                task_type = 'cpu' if i % 2 == 0 else 'io'
                future = executor.submit(mixed_task, i, task_type)
                tasks.append((i, future))
            
            for task_id, future in tasks:
                try:
                    result = future.result()
                    print(f"任务 {task_id} 完成: {result}")
                except Exception as e:
                    print(f"任务 {task_id} 执行出错: {e}")
    
    def custom_thread_pool_example(self):
        """自定义线程池示例（使用ThreadPoolExecutor的更多参数）"""
        print(f"\n=== 自定义线程池示例 ===")
        
        # 创建自定义线程池
        executor = concurrent.futures.ThreadPoolExecutor(
            max_workers=self.cpu_cores * 2,
            thread_name_prefix="CustomPool"
        )
        
        def task_with_name(task_id):
            thread_name = threading.current_thread().name
            print(f"任务 {task_id} 在线程 {thread_name} 中执行")
            time.sleep(0.5)
            return f"任务 {task_id} 完成"
        
        # 提交任务
        futures = [executor.submit(task_with_name, i) for i in range(10)]
        
        # 获取结果
        for future in concurrent.futures.as_completed(futures):
            try:
                result = future.result()
                print(result)
            except Exception as e:
                print(f"任务执行出错: {e}")
        
        # 关闭线程池
        executor.shutdown(wait=True)
    
    def process_pool_example(self):
        """进程池示例（适合CPU密集型任务）"""
        print(f"\n=== 进程池示例（CPU密集型）===")
        print(f"进程池大小: {self.cpu_cores}")
        
        # 对于CPU密集型任务，使用进程池可能更合适（避免GIL限制）
        with concurrent.futures.ProcessPoolExecutor(max_workers=self.cpu_cores) as executor:
            futures = [
                executor.submit(self.cpu_intensive_task, 1000000)
                for _ in range(10)
            ]
            
            for i, future in enumerate(concurrent.futures.as_completed(futures)):
                try:
                    result = future.result()
                    print(f"进程任务 {i+1} 完成，结果: {result}")
                except Exception as e:
                    print(f"任务执行出错: {e}")


def main():
    """主函数"""
    print("=" * 50)
    print("线程池线程数设置示例")
    print("=" * 50)
    
    example = ThreadPoolExample()
    
    # 运行各种示例
    example.cpu_intensive_example()
    example.io_intensive_example()
    example.mixed_task_example()
    example.custom_thread_pool_example()
    example.process_pool_example()
    
    print("\n" + "=" * 50)
    print("所有示例执行完成")
    print("=" * 50)


if __name__ == "__main__":
    main()

File: ThreadPoolExample.java
import java.util.concurrent.*;

/**
 * 线程池线程数设置示例
 * 
 * 线程池大小的设置原则：
 * 1. CPU密集型任务：线程数 = CPU核心数 + 1
 * 2. IO密集型任务：线程数 = CPU核心数 * (1 + IO等待时间/CPU计算时间)
 * 3. 混合型任务：需要根据实际情况调整
 */
public class ThreadPoolExample {
    
    // 获取CPU核心数
    private static final int CPU_CORES = Runtime.getRuntime().availableProcessors();
    
    /**
     * CPU密集型任务示例
     * 适合计算、加密、压缩等CPU占用高的任务
     */
    public static void cpuIntensiveExample() {
        // CPU密集型：线程数 = CPU核心数 + 1
        int threadPoolSize = CPU_CORES + 1;
        
        ThreadPoolExecutor executor = new ThreadPoolExecutor(
            threadPoolSize,           // 核心线程数
            threadPoolSize,           // 最大线程数
            60L,                      // 空闲线程存活时间
            TimeUnit.SECONDS,
            new LinkedBlockingQueue<>(100),  // 工作队列
            new ThreadFactory() {
                private int counter = 0;
                @Override
                public Thread newThread(Runnable r) {
                    return new Thread(r, "CPU-Task-" + (++counter));
                }
            },
            new ThreadPoolExecutor.CallerRunsPolicy()  // 拒绝策略
        );
        
        System.out.println("CPU密集型线程池大小: " + threadPoolSize);
        System.out.println("CPU核心数: " + CPU_CORES);
        
        // 提交CPU密集型任务
        for (int i = 0; i < 10; i++) {
            final int taskId = i;
            executor.submit(() -> {
                // CPU密集型计算
                long sum = 0;
                for (int j = 0; j < 1000000; j++) {
                    sum += j * j;
                }
                System.out.println("任务 " + taskId + " 完成，结果: " + sum);
            });
        }
        
        executor.shutdown();
        try {
            executor.awaitTermination(30, TimeUnit.SECONDS);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }
    
    /**
     * IO密集型任务示例
     * 适合网络请求、文件读写、数据库查询等IO操作
     */
    public static void ioIntensiveExample() {
        // IO密集型：线程数 = CPU核心数 * (1 + IO等待时间/CPU计算时间)
        // 通常IO等待时间远大于CPU计算时间，可以设置为 CPU核心数 * 2 到 CPU核心数 * 4
        int threadPoolSize = CPU_CORES * 2;
        
        ThreadPoolExecutor executor = new ThreadPoolExecutor(
            threadPoolSize,           // 核心线程数
            threadPoolSize * 2,       // 最大线程数（IO密集型可以更大）
            60L,
            TimeUnit.SECONDS,
            new LinkedBlockingQueue<>(200),
            new ThreadFactory() {
                private int counter = 0;
                @Override
                public Thread newThread(Runnable r) {
                    return new Thread(r, "IO-Task-" + (++counter));
                }
            },
            new ThreadPoolExecutor.CallerRunsPolicy()
        );
        
        System.out.println("IO密集型线程池大小: " + threadPoolSize);
        System.out.println("CPU核心数: " + CPU_CORES);
        
        // 提交IO密集型任务
        for (int i = 0; i < 20; i++) {
            final int taskId = i;
            executor.submit(() -> {
                try {
                    // 模拟IO操作（网络请求、文件读写等）
                    Thread.sleep(1000);  // 模拟IO等待
                    System.out.println("IO任务 " + taskId + " 完成");
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                }
            });
        }
        
        executor.shutdown();
        try {
            executor.awaitTermination(60, TimeUnit.SECONDS);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }
    
    /**
     * 使用Executors工具类创建线程池（不推荐，但简单）
     * 注意：Executors.newFixedThreadPool() 创建的线程池队列是无界的，可能导致OOM
     */
    public static void simpleExample() {
        // 固定大小线程池
        ExecutorService fixedPool = Executors.newFixedThreadPool(CPU_CORES);
        
        // 缓存线程池（适合大量短时任务）
        ExecutorService cachedPool = Executors.newCachedThreadPool();
        
        // 单线程池（保证任务顺序执行）
        ExecutorService singlePool = Executors.newSingleThreadExecutor();
        
        // 定时任务线程池
        ScheduledExecutorService scheduledPool = Executors.newScheduledThreadPool(CPU_CORES);
        
        // 使用示例
        fixedPool.submit(() -> System.out.println("固定线程池任务"));
        
        fixedPool.shutdown();
        cachedPool.shutdown();
        singlePool.shutdown();
        scheduledPool.shutdown();
    }
    
    /**
     * 动态调整线程池大小示例
     */
    public static void dynamicSizeExample() {
        ThreadPoolExecutor executor = new ThreadPoolExecutor(
            2,                        // 最小核心线程数
            10,                       // 最大线程数
            60L,
            TimeUnit.SECONDS,
            new LinkedBlockingQueue<>(50),
            new ThreadPoolExecutor.CallerRunsPolicy()
        );
        
        // 根据系统负载动态调整
        // 监控队列长度和活跃线程数
        executor.submit(() -> {
            while (!Thread.currentThread().isInterrupted()) {
                int queueSize = executor.getQueue().size();
                int activeCount = executor.getActiveCount();
                int poolSize = executor.getPoolSize();
                
                System.out.printf("队列大小: %d, 活跃线程: %d, 线程池大小: %d%n", 
                    queueSize, activeCount, poolSize);
                
                // 如果队列积压严重，可以考虑增加线程数
                if (queueSize > 40 && poolSize < executor.getMaximumPoolSize()) {
                    executor.setCorePoolSize(Math.min(poolSize + 2, executor.getMaximumPoolSize()));
                }
                
                try {
                    Thread.sleep(5000);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    break;
                }
            }
        });
        
        executor.shutdown();
    }
    
    public static void main(String[] args) {
        System.out.println("=== 线程池线程数设置示例 ===\n");
        
        System.out.println("1. CPU密集型任务示例:");
        cpuIntensiveExample();
        
        System.out.println("\n2. IO密集型任务示例:");
        ioIntensiveExample();
        
        System.out.println("\n3. 简单示例:");
        simpleExample();
        
        System.out.println("\n4. 动态调整示例:");
        dynamicSizeExample();
    }
}
